File: backend/controllers/users.controller.ts
import { hash } from 'bcrypt';
import BaseController from './BaseController';
import { UsersManager, ConnectionsLogsManager, DocumentsManager } from '../classes';
import { DEFAULT_DOCUMENT } from '../utils';

import type { Request, Response } from 'express';

export default class Authentification extends BaseController<UsersManager> {
  connections: ConnectionsLogsManager;
  documents: DocumentsManager;
  constructor(app: App) {
    super(new UsersManager(app));
    this.connections = new ConnectionsLogsManager(app);
    this.documents = new DocumentsManager(app);
  }
  public async getMe(req: Request, res: Response) {
    try {
      const user = await this.manager.getPublic(req.user_id as string);
      const last_connection = await this.connections.getLastConnection(req.user_id as string);
      res.status(200).json(this.utils.success({ user, last_connection }));
    } catch (e) {
      res.status(500).json(this.utils.error(e));
    }
  }
  public async getById(req: Request, res: Response) {
    try {
      const user = await this.manager.getPublic(req.params.id as string);
      const last_connection = await this.connections.getLastConnection(req.params.id as string);
      res.status(200).json(this.utils.success({ user, last_connection }));
    } catch (e) {
      res.status(500).json(this.utils.error(e));
    }
  }
  public async getAll(_: Request, res: Response) {
    try {
      const users = await this.manager.getAll();
      res.status(200).json(this.utils.success({ users }));
    } catch (e) {
      res.status(500).json(this.utils.error(e));
    }
  }
  public async post(req: Request, res: Response) {
    try {
      const passwordStr = req.body.password as string;
      if (!passwordStr) return res.status(400).json(this.utils.error('Password must be provided.'));
      const hashed = await hash(passwordStr, 10);
      const user = await this.manager.exists(req.body.username || '');
      if (user) return res.status(400).json(this.utils.error('Username already taken.'));
      const data = {
        id: this.app.snowflake.generate().toString(),
        username: req.body.username as string,
        firstname: req.body.firstname as string,
        lastname: req.body.lastname as string,
        role: 1,
        avatar: req.body.avatar as string,
        email: req.body.email as string,
        password: hashed,
        created_timestamp: Date.now().toString(),
        updated_timestamp: Date.now().toString(),
      };
      const error = this.manager.validator.validate(data);
      if (error) return res.status(400).json(this.utils.error(error));
      const r = await this.manager.post(data);
      await this.documents.add({
        id: this.app.snowflake.generate().toString(),
        author_id: data.id,
        created_timestamp: Date.now().toString(),
        updated_timestamp: Date.now().toString(),
        ...DEFAULT_DOCUMENT,
      });
      return res.status(201).json(this.utils.success(r));
    } catch (e) {
      return res.status(500).json(this.utils.error(e));
    }
  }
  public async update(req: Request, res: Response) {
    const user_db = await this.manager.getById(req.params.id!);
    if (!user_db) return res.status(404).json(this.utils.error('User not found.'));
    if (req.params.id != req.user_id && req.user_role !== 2) return res.status(401).json(this.utils.error('Unauthorized'));
    return this.manager
      .put(req.params.id!, {
        username: user_db.username,
        firstname: req.body.firstname ?? user_db.firstname,
        lastname: req.body.lastname ?? user_db.lastname,
        role: user_db.role,
        avatar: req.body.avatar ?? user_db.avatar,
        email: req.body.email ?? user_db.email,
        updated_timestamp: Date.now().toString(),
      })
      .then(r => res.status(200).json(this.utils.success(r)))
      .catch(e => res.status(500).json(this.utils.error(e)));
  }
  public async updatePassword(req: Request, res: Response) {
    try {
      const passwordStr = req.body.password as string;
      if (!passwordStr) return res.status(400).json(this.utils.error('Password must be provided.'));
      if (req.params.id != req.user_id) return res.status(401).json(this.utils.error('Unauthorized'));
      const hashed = await hash(passwordStr, 10);
      const r = await this.manager.updatePassword({
        id: req.params.id!,
        password: hashed,
      });
      return res.status(200).json(this.utils.success(r));
    } catch (e) {
      return res.status(500).json(this.utils.error(e));
    }
  }
}

File: backend/controllers/users.controller.go
package controllers

import (
	"alexandrie/app"
	"alexandrie/models"
	"alexandrie/utils"
	"errors"
	"net/http"
	"os"

	"github.com/gin-gonic/gin"
)

type UserController interface {
	GetUsers(c *gin.Context) (int, any)
	GetUserById(c *gin.Context) (int, any)
	GetPublicUser(c *gin.Context) (int, any)
	CreateUser(c *gin.Context) (int, any)
	UpdateUser(c *gin.Context) (int, any)
	UpdatePassword(c *gin.Context) (int, any)
	DeleteUser(c *gin.Context) (int, any)
	AdminUpdatePassword(c *gin.Context) (int, any)
	GetUserSessions(c *gin.Context) (int, any)
	DeleteUserSession(c *gin.Context) (int, any)
	SuspendUser(c *gin.Context) (int, any)
}

func NewUserController(app *app.App) UserController {
	return &Controller{
		app: app,
	}
}

// Get Users
// @Summary Get all users
// @Method GET
// @Router /users [get]
// @Security Authenfification: Auth, Admin
// @Success 200 {object} Success([]models.User)
// @Failure 400 {object} Error
// @Failure 401 {object} Error
func (ctr *Controller) GetUsers(c *gin.Context) (int, any) {
	users, err := ctr.app.Services.User.GetAllUsers()
	if err != nil {
		return http.StatusInternalServerError, err
	}
	return http.StatusOK, users
}

// Get public user profile by username or email
// @Summary Get public user profile by username or email
// @Method GET
// @Router /users/public/{usernameOrEmail} [get]
// @Security None
func (ctr *Controller) GetPublicUser(c *gin.Context) (int, any) {
	usernameOrEmail := c.Param("query")
	if usernameOrEmail == "" {
		return http.StatusBadRequest, errors.New("username or email is required")
	}
	users, err := ctr.app.Services.User.SearchPublicUsers(usernameOrEmail)
	if err != nil {
		return http.StatusInternalServerError, err
	}
	if len(users) == 0 {
		return http.StatusNotFound, errors.New("user not found")
	}

	return http.StatusOK, users
}

// Get User by ID
// @Summary Get user by ID
// @Method GET
// @Router /users/{id} [get]
// @Security Authenfification: Auth, Admin
// @Param id path int true "User ID"
// @Success 200 {object} Success(models.User)
// @Failure 400 {object} Error
// @Failure 401 {object} Error
func (ctr *Controller) GetUserById(c *gin.Context) (int, any) {
	if _, err := actorFromRequest(c); err != nil {
		return statusFromAccessError(err), err
	}

	targetUserId, err := utils.GetTargetId(c, c.Param("userId"))
	if err != nil {
		return http.StatusBadRequest, err
	}

	result, err := ctr.app.Services.User.GetUserById(c.Request.Context(), targetUserId)
	if err != nil {
		return statusFromAccessError(err), err
	}

	return http.StatusOK, result
}

// Create User
// @Summary Create a new user
// @Method POST
// @Router /users [post]
// @Security None
// @Body Username*, Firstname, Lastname, Avatar, Role, Email*, Password*
// @Success 201 {object} Success(models.User)
// @Failure 400 {object} Error
func (ctr *Controller) CreateUser(c *gin.Context) (int, any) {

	disabled := os.Getenv("CONFIG_DISABLE_SIGNUP")

	if disabled == "true" {
		return http.StatusForbidden, errors.New("user signup is disabled")
	}

	var user models.User
	if err := c.ShouldBind(&user); err != nil {
		return http.StatusBadRequest, err
	}

	firstname := utils.StringValue(user.Firstname)
	lastname := utils.StringValue(user.Lastname)
	avatar := utils.StringValue(user.Avatar)
	email := utils.StringValue(user.Email)
	password := utils.StringValue(user.Password)
	user_type := user.Type
	totp_forced := user.TOTPForced

	createdUser, err := ctr.app.Services.User.CreateUser(user.Username, firstname, lastname, avatar, email, user_type, password, totp_forced)
	if err != nil {
		return http.StatusBadRequest, err
	}

	return http.StatusCreated, createdUser
}

// Update User
// @Summary Update user information
// @Method PATCH
// @Router /users/{id} [patch]
// @Security Authenfification: Auth
// @Param id path int true "User ID"
// @Body Username, Firstname, Lastname, Avatar, Role, Email
// @Success 200 {object} Success(models.User)
// @Failure 400 {object} Error
// @Failure 401 {object} Error
func (ctr *Controller) UpdateUser(c *gin.Context) (int, any) {
	if _, err := actorFromRequest(c); err != nil {
		return statusFromAccessError(err), err
	}

	targetUserId, err := utils.GetTargetId(c, c.Param("userId"))
	if err != nil {
		return http.StatusBadRequest, err
	}

	var user models.User
	if err := c.ShouldBind(&user); err != nil {
		return http.StatusBadRequest, err
	}

	updatedUser, err := ctr.app.Services.User.UpdateUser(c.Request.Context(), targetUserId, user.Firstname, user.Lastname, user.Avatar, user.Email, user.Type, user.TOTPForced)
	if err != nil {
		return statusFromAccessError(err), err
	}
	return http.StatusOK, updatedUser
}

// Update Password
// @Summary Update user password
// @Method PATCH
// @Router /users/{id}/password [patch]
// @Security Authenfification: Auth, {self, admin}
// @Param id path int true "User ID"
// @Body CurrentPassword, Password
// @Success 200 {object} Success(string)
// @Failure 400 {object} Error
// @Failure 401 {object} Error
func (ctr *Controller) UpdatePassword(c *gin.Context) (int, any) {
	if _, err := actorFromRequest(c); err != nil {
		return statusFromAccessError(err), err
	}

	targetUserId, err := utils.GetTargetId(c, c.Param("userId"))
	if err != nil {
		return http.StatusBadRequest, err
	}

	var payload struct {
		CurrentPassword string `form:"current_password" json:"current_password"`
		Password        string `form:"password" json:"password"`
	}
	if err := c.ShouldBind(&payload); err != nil {
		return http.StatusBadRequest, errors.New("invalid request payload")
	}

	err = ctr.app.Services.User.UpdatePassword(c.Request.Context(), targetUserId, payload.CurrentPassword, payload.Password)
	if err != nil {
		return statusFromAccessError(err), err
	}
	return http.StatusOK, "Password updated successfully"
}

// Delete User
// @Summary Delete user
// @Method DELETE
// @Router /users/{id} [delete]
// @Security Authenfification: Auth, {self, admin}
// @Param id path int true "User ID"
// @Success 200 {object} Success(string)
// @Failure 400 {object} Error
// @Failure 401 {object} Error
func (ctr *Controller) DeleteUser(c *gin.Context) (int, any) {
	if _, err := actorFromRequest(c); err != nil {
		return statusFromAccessError(err), err
	}

	targetUserId, err := utils.GetTargetId(c, c.Param("userId"))
	if err != nil {
		return http.StatusBadRequest, err
	}
	err = ctr.app.Services.User.DeleteUser(c.Request.Context(), targetUserId, ctr.app.Services.Minio)
	if err != nil {
		return statusFromAccessError(err), err
	}
	return http.StatusOK, "User deleted successfully"
}

func (ctr *Controller) AdminUpdatePassword(c *gin.Context) (int, any) {
	targetUserId, err := utils.GetTargetId(c, c.Param("userId"))
	if err != nil {
		return http.StatusBadRequest, err
	}

	var payload struct {
		Password string `json:"password"`
	}
	if err := c.ShouldBindJSON(&payload); err != nil {
		return http.StatusBadRequest, err
	}

	err = ctr.app.Services.User.AdminUpdatePassword(c.Request.Context(), targetUserId, payload.Password)
	if err != nil {
		return http.StatusInternalServerError, err
	}
	return http.StatusOK, "Password updated successfully"
}

func (ctr *Controller) GetUserSessions(c *gin.Context) (int, any) {
	targetUserId, err := utils.GetTargetId(c, c.Param("userId"))
	if err != nil {
		return http.StatusBadRequest, err
	}

	sessions, err := ctr.app.Services.Session.GetSessionsByUserId(targetUserId)
	if err != nil {
		return http.StatusInternalServerError, err
	}
	return http.StatusOK, sessions
}

func (ctr *Controller) DeleteUserSession(c *gin.Context) (int, any) {
	targetUserId, err := utils.GetTargetId(c, c.Param("userId"))
	if err != nil {
		return http.StatusBadRequest, err
	}

	sessionId, err := utils.GetTargetId(c, c.Param("sessionId"))
	if err != nil {
		return http.StatusBadRequest, err
	}

	err = ctr.app.Services.Session.DeleteSession(sessionId, targetUserId)
	if err != nil {
		return http.StatusInternalServerError, err
	}
	return http.StatusOK, "Session deleted successfully"
}

func (ctr *Controller) SuspendUser(c *gin.Context) (int, any) {
	targetUserId, err := utils.GetTargetId(c, c.Param("userId"))
	if err != nil {
		return http.StatusBadRequest, err
	}

	var payload struct {
		Suspend bool `json:"suspend"`
	}
	if err := c.ShouldBindJSON(&payload); err != nil {
		return http.StatusBadRequest, err
	}

	err = ctr.app.Services.User.SuspendUser(c.Request.Context(), targetUserId, payload.Suspend)
	if err != nil {
		return http.StatusInternalServerError, err
	}

	if payload.Suspend {
		err = ctr.app.Services.Session.DeleteAllByUser(targetUserId)
		if err != nil {
			return http.StatusInternalServerError, err
		}
	}

	return http.StatusOK, "User suspended state updated successfully"
}
